File: ThreeNumberSum.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class ThreeNumberSum {
    public static List<Integer[]> threeNumberSum(int[] array, int targetSum){
        Arrays.sort(array);
        int currentSum = 0;
        List<Integer[]> al = new ArrayList<>();
        for(int i = 0;i< array.length;i++){
            int left = i+1;
            int right = array.length-1;
            while(left< right){
                currentSum = array[i] + array[left] + array[right];
                if(currentSum == targetSum){
                    Integer[] triplets = {array[i],array[left],array[right]};
                    al.add(triplets);
                    left ++;
                    right--;
                }
                else if(currentSum < targetSum){
                    left ++;
                }
                else{
                    right--;
                }
            }
        }
        return al;
    }
}

File: ThreeNumberSum.js
//Time Complexity O(n^2) | Space Complexity O(n)
function ThreeNumberSum(array, targetSum) {
  array.sort();
  const triplets = [];
  for (let i = 0; i < array.length - 2; i++) {
    let left = i + 1;
    let right = array.length - 1;
    while (left < right) {
      const currentSum = array[i] + array[left] + array[right];
      if (currentSum === targetSum) {
        triplets.push([array[i], array[left], array[right]]);
        left++;
        right--;
      } else if (currentSum < targetSum) {
        left++;
      } else {
        right--;
      }
    }
  }
  return triplets;
}

const result = ThreeNumberSum([12, 3, 1, 2, -6, 5, -8, 6], 0);
console.log(result[0][2]);
